runtime/auth/aws-signing-common/common/src/aws/smithy/kotlin/runtime/auth/awssigning/Presigner.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+package aws.smithy.kotlin.runtime.auth.awssigning
+
+import aws.smithy.kotlin.runtime.auth.awscredentials.CredentialsProvider
+import aws.smithy.kotlin.runtime.http.*
+import aws.smithy.kotlin.runtime.http.endpoints.Endpoint
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.util.InternalApi
+import kotlin.time.Duration
+
+data class SigningContext(val service: String?, val region: String?)
+data class SigningContextualizedEndpoint(val endpoint: Endpoint, val context: SigningContext?)
+typealias SigningEndpointProvider = suspend (SigningContext) -> SigningContextualizedEndpoint
```

Comment:
this could be a functional sam interface:

```kotlin
fun interface SigningEndpointProvider {
    suspend fun resolve(context: SigningContext): SigningContextualizedEndpoint
}
```

This allows you to pass lambdas in or implementations of the interface

Reply:
This is already a lambda type so it does accept lambdas as values. See its [usage in `AwsEndpointResolver`](https://github.com/awslabs/aws-sdk-kotlin/blob/fce177b205be09ec6b8d297f8cd1427b12a45336/aws-runtime/aws-endpoint/common/src/aws/sdk/kotlin/runtime/endpoint/AwsEndpointResolver.kt#L26-L29). Converting it a SAM interface would _also_ allow traditional subclass implementations but would not shorten the syntax for using lambda instances.

Reply:
Right I was more saying is there any advantage to a typealias over a SAM interface? I don't have a strong opinion here though

Reply:
The advantage I see is declaration brevity and the absence of a formal method name we never intend to use.